File: api/app/Models/Showcase.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

/**
 * "Showcase"/"Onyesho" (CLAUDE.md Part 3) — a vertical full-screen product
 * video. `product_id` is non-nullable at the schema level (see the migration).
 */
#[Fillable(['seller_id', 'product_id', 'video_path', 'thumb_path', 'caption', 'duration'])]
class Showcase extends Model
{
    use HasFactory;

    protected function casts(): array
    {
        return [
            'views' => 'integer',
        ];
    }

    public function seller(): BelongsTo
    {
        return $this->belongsTo(SellerProfile::class, 'seller_id');
    }

    public function product(): BelongsTo
    {
        return $this->belongsTo(Product::class);
    }

    /** Publicly visible: seller verified, Listing still visible. */
    public function scopeVisible(Builder $query): Builder
    {
        return $query
            ->whereHas('seller', fn (Builder $q) => $q->where('status', 'verified'))
            ->whereHas('product', fn (Builder $q) => $q->where('is_active', true)->where('is_hidden', false));
    }

    public function scopeFollowedBy(Builder $query, int $buyerId): Builder
    {
        return $query->whereHas('seller', fn (Builder $q) => $q->whereHas(
            'followers',
            fn (Builder $f) => $f->where('users.id', $buyerId)
        ));
    }
}

File: app/lib/data/models/showcase.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'product_summary.dart';
import 'seller_summary.dart';

part 'showcase.freezed.dart';
part 'showcase.g.dart';

/// "Showcase"/"Onyesho" (CLAUDE.md Part 3) — a vertical full-screen product
/// video. Mirrors `App\Http\Resources\ShowcaseResource`.
@freezed
abstract class Showcase with _$Showcase {
  const factory Showcase({
    required int id,
    @JsonKey(name: 'video_path') required String videoPath,
    @JsonKey(name: 'thumb_path') required String thumbPath,
    String? caption,
    required int duration,
    @Default(0) int views,
    SellerSummary? seller,
    ProductSummary? product,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _Showcase;

  factory Showcase.fromJson(Map<String, dynamic> json) => _$ShowcaseFromJson(json);
}
